File: sga/individual.py
import numpy as np
import random

from typing import Callable, Iterable, List, Tuple, Optional

from .config import Config


class Individual:
    def __init__(self, config: Config, activation_function: Callable, out_activation_function: Callable):
        self.config = config

        # Variables initialization
        self.n_input = config.IND_INP_NUMBER
        self.n_output = config.IND_OUT_NUMBER
        self.max_node = config.IND_MAX_NODES
        self.total_size = self.n_input + self.n_output + self.max_node
        self.activation_function = activation_function
        self.out_activation_function = out_activation_function

        # Genes generation
        self.genes = np.full((self.total_size, self.total_size), np.inf)
        self.values = np.full(self.total_size, np.inf)

        if config.INITIAL_GENERATION == "full":
            for i in range(self.n_input + self.max_node, self.total_size):
                for j in range(self.n_input):
                    self.genes[i, j] = random.uniform(-1, 1)
        elif config.INITIAL_GENERATION == "random":
            for i in range(self.n_input + self.max_node, self.total_size):
                for j in range(self.n_input):
                    if random.random() < 0.5:
                        self.genes[i, j] = random.uniform(-1, 1)
        elif config.INITIAL_GENERATION == "random-nodes":
            for i in range(self.n_input + self.max_node, self.total_size):
                for j in range(self.n_input):
                    self.genes[i, j] = random.uniform(-1, 1)
            for _ in range(self.n_input, self.n_input + self.max_node):
                if random.random() < 0.5:
                    self.add_node()
            for _ in range(self.n_input, self.n_input + self.max_node):
                if random.random() < 0.5:
                    self.add_gene()

    def __available_genes(self) -> List[Tuple[int, int]]:
        """ Returns the list of available genes """
        available_genes = []
        for i in self.__valid_to_nodes():
            for j in self.__valid_from_nodes():
                if self.genes[i, j] == np.inf and self.genes[j, i] == np.inf and i != j:
                    available_genes.append((i, j))
        return available_genes

    def __existing_genes(self) -> List[Tuple[int, int]]:
        """ Returns the list of existing genes """
        existing_genes = []
        for i in self.__valid_to_nodes():
            for j in self.__valid_from_nodes():
                if self.genes[i, j] != np.inf:
                    existing_genes.append((i, j))
        return existing_genes

    def __valid_from_nodes(self) -> List[int]:
        """ Returns the list of nodes usable as input """
        valid_nodes = list(range(self.n_input))
        for i in range(self.n_input, self.n_input + self.max_node):
            if not np.all(self.genes[i] == np.inf) and not np.all(self.genes[:, i] == np.inf):
                valid_nodes.append(i)
        return valid_nodes

    def __valid_to_nodes(self) -> List[int]:
        """ Returns the list of nodes usable as output """
        valid_nodes = list(range(self.n_input + self.max_node, self.total_size))
        for i in range(self.n_input, self.n_input + self.max_node):
            if not np.all(self.genes[i] == np.inf) and not np.all(self.genes[:, i] == np.inf):
                valid_nodes.append(i)
        return valid_nodes

    def __check_path(self, node: int, already_checked: Optional[List[int]] = None):
        """ Checks the network, and removes undesirables genes """
        if already_checked is None:
            already_checked = []
        if not node < self.n_input:
            if node in already_checked:
                self.genes[already_checked[-1], node] = np.inf

                if np.all(self.genes[:, node] == np.inf):
                    for i in range(self.n_input, self.n_input + self.max_node):
                        if self.genes[node, i] != np.inf:
                            self.genes[node, i] = np.inf
                            self.__check_path(i)

            if np.all(self.genes[node] == np.inf):
                for i in range(self.n_input, self.total_size):
                    if self.genes[i, node] != np.inf:
                        self.genes[i, node] = np.inf
                        self.__check_path(i)

            for i in range(self.total_size):
                if self.genes[node, i] != np.inf:
                    self.__check_path(i, already_checked + [node])

    def __process(self, node: int):
        """ Predicts the node (i) """
        out = 0
        for i in range(self.n_input + self.max_node):
            if self.genes[node, i] != np.inf:
                if self.values[i] == np.inf:
                    self.__process(i)
                out += self.values[i] * self.genes[node, i]
        if node < self.n_input + self.max_node:
            self.values[node] = self.activation_function(out)
        else:
            self.values[node] = self.out_activation_function(out)

    def add_gene(self):
        """ Adds a new gene between two existing nodes """
        available_genes = self.__available_genes()
        if available_genes:
            new_gene = random.choice(available_genes)
            self.genes[new_gene[0], new_gene[1]] = random.uniform(- self.config.WEIGHT_AMP, self.config.WEIGHT_AMP)

    def add_node(self):
        """ Adds a new node in the middle of a gene """
        existing_genes = self.__existing_genes()
        if not existing_genes:
            self.add_gene()
            return

        empty_nodes = []
        for i in range(self.n_input, self.n_input + self.max_node):
            if np.all(self.genes[i] == np.inf):
                empty_nodes.append(i)

        if empty_nodes:
            rep_gene = random.choice(existing_genes)
            new_node = random.choice(empty_nodes)
            self.genes[rep_gene[0], rep_gene[1]] = np.inf
            self.genes[rep_gene[0], new_node] = random.uniform(- self.config.WEIGHT_AMP, self.config.WEIGHT_AMP)
            self.genes[new_node, rep_gene[1]] = random.uniform(- self.config.WEIGHT_AMP, self.config.WEIGHT_AMP)

    def remove_gene(self):
        """ Removes an existing gene """
        existing_genes = self.__existing_genes()
        if len(existing_genes) > 1:
            selected_gene = random.choice(existing_genes)
            self.genes[selected_gene[0], selected_gene[1]] = np.inf

    def remove_node(self):
        """ Removes a node and rebuilds the network partially """
        existing_nodes = []
        for i in range(self.n_input, self.n_input + self.max_node):
            if np.any(self.genes[i] != np.inf):
                existing_nodes.append(i)

        if existing_nodes:
            rem_node = random.choice(existing_nodes)
            from_nodes, to_nodes = [], []
            for i in range(self.n_input + self.max_node):
                if self.genes[rem_node, i] != np.inf:
                    from_nodes.append(i)
                    self.genes[rem_node, i] = np.inf
            for i in range(self.n_input, self.total_size):
                if self.genes[i, rem_node] != np.inf:
                    to_nodes.append(i)
                    self.genes[i, rem_node] = np.inf

            for i in to_nodes:
                for j in from_nodes:
                    if self.genes[i, j] == np.inf and random.random() < 0.5:
                        self.genes[i, j] = random.uniform(- self.config.WEIGHT_AMP, self.config.WEIGHT_AMP)

    def mutate(self, rate: float, amplitude: float):
        """ Mutates the genes """
        for i in range(self.n_input, self.total_size):
            for j in range(self.n_input + self.max_node):
                if random.random() < rate:
                    if self.genes[i, j] != np.inf:
                        self.genes[i, j] += random.uniform(- amplitude, amplitude)

    def cleanup(self):
        """ Cleans the network by removing undesirables genes """
        for i in range(self.n_input + self.max_node, self.total_size):
            self.__check_path(i)

    def predict(self, inputs: Iterable[float]) -> Iterable:
        """ Returns the prediction of the network """
        self.values.fill(np.inf)
        self.values[:self.n_input] = inputs
        for i in range(self.n_input + self.max_node, self.total_size):
            self.__process(i)
        return self.values[self.n_input + self.max_node:self.total_size]

File: src/main/java/com/github/gmattis/simplegeneticalgorithm/beans/algo/Individual.java
package com.github.gmattis.simplegeneticalgorithm.beans.algo;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Random;

import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.cpu.nativecpu.NDArray;
import org.nd4j.linalg.factory.Nd4j;
import org.nd4j.linalg.indexing.INDArrayIndex;
import org.nd4j.linalg.indexing.NDArrayIndex;

import com.github.gmattis.simplegeneticalgorithm.beans.config.Config;
import com.github.gmattis.simplegeneticalgorithm.beans.config.IndividualConfig;
import com.github.gmattis.simplegeneticalgorithm.beans.config.MutationsConfig;
import com.github.gmattis.simplegeneticalgorithm.util.Activation;
import com.github.gmattis.simplegeneticalgorithm.util.Util;

public class Individual {

	private IndividualConfig indivConfig;
	private MutationsConfig mutConfig;

	private int nInput;
	private int nOutput;
	private int maxNode;
	private int totalSize;
	private Activation activation;
	private Activation outActivation;

	private INDArray genes;
	private INDArray values;
	private Random rdm;

	public Individual(Config config, Activation activation, Activation outActivation,
			boolean generateNetwork) {
		this.rdm = new Random();
		//rdm.setSeed(System.currentTimeMillis());
		
		this.indivConfig = config.getIndividual();
		this.mutConfig = config.getMutations();

		// Variables init
		this.nInput = indivConfig.getInputNumber();
		this.nOutput = indivConfig.getOutputNumber();
		this.maxNode = indivConfig.getMaxNodes();
		this.totalSize = nInput + nOutput + maxNode;
		this.activation = activation;
		this.outActivation = outActivation;

		// Genes generation
		this.genes = Nd4j.zeros(totalSize, totalSize).addi(Double.POSITIVE_INFINITY);
		this.values = Nd4j.zeros(totalSize).addi(Double.POSITIVE_INFINITY);

		if (generateNetwork) {
			switch (indivConfig.getInitialGeneration()) {
			case FULL:
				for (int i = nInput + maxNode; i < totalSize; i++) {
					for (int j = 0; j < nInput; j++) {
						genes.putScalar(i, j, 2. * rdm.nextDouble() - 1.);
					}
				}
				break;
			case RANDOM:
				for (int i = nInput + maxNode; i < totalSize; i++) {
					for (int j = 0; j < nInput; j++) {
						if (rdm.nextDouble() < 0.5) {
							genes.putScalar(i, j, 2. * rdm.nextDouble() - 1.);
						}
					}
				}
				break;
			case RANDOM_NODES:
				for (int i = nInput + maxNode; i < totalSize; i++) {
					for (int j = 0; j < nInput; j++) {
						genes.putScalar(i, j, 2 * rdm.nextDouble() - 1);
					}
				}

				for (int i = nInput; i < nInput + maxNode; i++) {
					if (rdm.nextDouble() < 0.5) {
						this.addNode();
					}
				}

				for (int i = nInput; i < nInput + maxNode; i++) {
					if (rdm.nextDouble() < 0.5) {
						this.addGene();
					}
				}
				break;
			default:
				throw new IllegalArgumentException("Initial Generation must be `FULL`, `RANDOM`, or `RANDOM_NODES`");
			}
		}
	}

	private List<int[]> availableGenes() {
		List<int[]> availableGenes = new ArrayList<int[]>();
		for (int i : validToNodes()) {
			for (int j : validFromNodes()) {
				if (genes.getDouble(i, j) == Double.POSITIVE_INFINITY
						&& genes.getDouble(j, i) == Double.POSITIVE_INFINITY && j != i) {
					availableGenes.add(new int[] { i, j });
				}
			}
		}
		return availableGenes;
	}

	private List<int[]> existingGenes() {
		List<int[]> existingGenes = new ArrayList<int[]>();
		for (int i : validToNodes()) {
			for (int j : validFromNodes()) {
				if (genes.getDouble(i, j) != Double.POSITIVE_INFINITY) {
					existingGenes.add(new int[] { i, j });
				}
			}
		}
		return existingGenes;
	}

	private List<Integer> validFromNodes() {
		List<Integer> validNodes = Util.listRange(0, nInput);
		for (int i = nInput; i < nInput + maxNode; i++) {
			if (!Util.all(genes.slice(i, 0).reshape(totalSize), Double.POSITIVE_INFINITY)
					&& !Util.all(genes.slice(i, 1).reshape(totalSize), Double.POSITIVE_INFINITY)) {
				validNodes.add(i);
			}
		}
		return validNodes;
	}

	private List<Integer> validToNodes() {
		List<Integer> validNodes = Util.listRange(nInput + maxNode, totalSize);
		for (int i = nInput; i < nInput + maxNode; i++) {
			if (!Util.all(genes.slice(i, 0).reshape(totalSize), Double.POSITIVE_INFINITY)
					&& !Util.all(genes.slice(i, 1).reshape(totalSize), Double.POSITIVE_INFINITY)) {
				validNodes.add(i);
			}
		}
		return validNodes;
	}

	private void checkPath(int node, List<Integer> alreadyChecked) {
		if (!(node < nInput)) {
			if (alreadyChecked.contains(node)) {
				genes.putScalar(alreadyChecked.get(alreadyChecked.size() - 1), node, Double.POSITIVE_INFINITY);

				if (Util.all(genes.slice(node, 1).reshape(totalSize), Double.POSITIVE_INFINITY)) {
					for (int i = nInput; i < nInput + maxNode; i++) {
						if (genes.getDouble(node, i) != Double.POSITIVE_INFINITY) {
							genes.putScalar(node, i, Double.POSITIVE_INFINITY);
							checkPath(i, new ArrayList<Integer>());
						}
					}
				}
			}

			if (Util.all(genes.slice(node, 0).reshape(totalSize), Double.POSITIVE_INFINITY)) {
				for (int i = nInput; i < totalSize; i++) {
					if (genes.getDouble(i, node) != Double.POSITIVE_INFINITY) {
						genes.putScalar(i, node, Double.POSITIVE_INFINITY);
						checkPath(i, new ArrayList<Integer>());
					}
				}
			}

			for (int i = 0; i < totalSize; i++) {
				if (genes.getDouble(node, i) != Double.POSITIVE_INFINITY) {
					List<Integer> pathCopy = new ArrayList<Integer>(alreadyChecked);
					pathCopy.add(i);
					checkPath(i, pathCopy);
				}
			}
		}
	}

	private void process(int node) {
		double out = 0.;
		for (int i = 0; i < nInput + maxNode; i++) {
			if (genes.getDouble(node, i) != Double.POSITIVE_INFINITY) {
				if (values.getDouble(i) == Double.POSITIVE_INFINITY) {
					process(i);
				}
				out += values.getDouble(i) * genes.getDouble(node, i);
			}
		}
		try {
			if (node < nInput + maxNode) {
				Method act = Activation.class.getMethod(activation.getName(), double.class);
				values.putScalar(node, (double) act.invoke(null, out));
			} else {
				Method act = Activation.class.getMethod(outActivation.getName(), double.class);
				values.putScalar(node, (double) act.invoke(null, out));
			}
		} catch (IllegalAccessException | IllegalArgumentException | InvocationTargetException | NoSuchMethodException | SecurityException e) {
			throw new InternalError();
		}
	}

	public void addGene() {
		List<int[]> availableGenes = availableGenes();
		if(!availableGenes.isEmpty()) {
			int[] newGene = Util.rdmChoice(rdm, availableGenes);
			genes.putScalar(newGene[0], newGene[1], mutConfig.getWeightsAmplitude()*(2*rdm.nextDouble() - 1));
		}
	}

	public void addNode() {
		List<int[]> existingGenes = existingGenes();
		if (existingGenes.isEmpty()) {
			addGene();
			return;
		}
		
		List<Integer> emptyNodes = new ArrayList<Integer>();
		for(int i = nInput; i < nInput + maxNode; i++) {
			if(Util.all(genes.slice(i, 0).reshape(totalSize), Double.POSITIVE_INFINITY)) {
				emptyNodes.add(i);
			}
		}
		
		if(!emptyNodes.isEmpty()) {
			int[] repGene = Util.rdmChoice(rdm, existingGenes);
			int newNode = Util.rdmChoice(rdm, emptyNodes);
			
			genes.putScalar(repGene[0], repGene[1], Double.POSITIVE_INFINITY);
			genes.putScalar(repGene[0], newNode, mutConfig.getWeightsAmplitude()*(2*rdm.nextDouble() - 1));
			genes.putScalar(newNode, repGene[1], mutConfig.getWeightsAmplitude()*(2*rdm.nextDouble() - 1));
		}
	}
	
	public void removeGene() {
		List<int[]> existingGenes = existingGenes();
		if(existingGenes.size() > 1) {
			int[] selectedGene = Util.rdmChoice(rdm, existingGenes);
			genes.putScalar(selectedGene[0], selectedGene[1], Double.POSITIVE_INFINITY);
		}
	}
	
	public void removeNodes() {
		List<Integer> existingNodes = new ArrayList<Integer>();
		for(int i = nInput; i < nInput + maxNode; i++) {
			if(!Util.all(genes.slice(i, 0).reshape(totalSize), Double.POSITIVE_INFINITY)) {
				existingNodes.add(i);
			}
		}
		
		if(!existingNodes.isEmpty()) {
			int remNode = Util.rdmChoice(rdm, existingNodes);
			List<Integer> fromNodes = new ArrayList<Integer>(), toNodes = new ArrayList<Integer>();
			
			for(int i = 0; i < nInput + maxNode; i++) {
				if(genes.getDouble(remNode, i) != Double.POSITIVE_INFINITY) {
					fromNodes.add(i);
					genes.put(remNode, i, Double.POSITIVE_INFINITY);
				}
			}
			
			for(int i = nInput; i < totalSize; i++) {
				if(genes.getDouble(i, remNode) != Double.POSITIVE_INFINITY) {
					toNodes.add(i);
					genes.put(i, remNode, Double.POSITIVE_INFINITY);
				}
			}
			
			for(int i : toNodes) {
				for(int j : fromNodes) {
					if (genes.getDouble(i, j) != Double.POSITIVE_INFINITY && rdm.nextDouble() < 0.5) {
						genes.putScalar(i, j, mutConfig.getWeightsAmplitude()*(2*rdm.nextDouble() - 1));
					}
				}
			}
		}
	}
	
	public void mutate(double rate, double amplitude) {
		for(int i = nInput; i < totalSize; i++) {
			for(int j = 0; j < nInput + maxNode; j++) {
				if(rdm.nextDouble() < rate && genes.getDouble(i, j) != Double.POSITIVE_INFINITY) {
					genes.putScalar(i, j, genes.getDouble(i, j) + amplitude*(2*rdm.nextDouble() - 1));
				}
			}
		}
	}
	
	public void cleanup() {
		for(int i = nInput + maxNode; i < totalSize; i++) {
			checkPath(i, new ArrayList<Integer>());
		}
	}
	
	public INDArray predict(INDArray inputs) {
		values.addi(Double.POSITIVE_INFINITY);
		INDArrayIndex[] indices = new INDArrayIndex[] {NDArrayIndex.interval(0, nInput)};
		values.put(indices, inputs);
		
		for(int i = nInput + maxNode; i < totalSize; i++) {
			process(i);
		}
		
		return values.get(NDArrayIndex.interval(nInput + maxNode, totalSize));
	}
}
